protofsm/state_machine_test.go

```diff
@@ -0,0 +1,512 @@
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+type dummyEvents interface {
+	dummy()
+}
+
+type goToFin struct {
+}
+
+func (g *goToFin) dummy() {
+}
+
+type emitInternal struct {
+}
+
+func (e *emitInternal) dummy() {
+}
+
+type daemonEvents struct {
+}
+
+func (s *daemonEvents) dummy() {
+}
+
+type dummyEnv struct {
+	mock.Mock
+}
+
+func (d *dummyEnv) CleanUp() error {
+	args := d.Called()
+
+	return args.Error(0)
+}
+
+func (d *dummyEnv) Name() string {
+	return "test"
+}
+
+type dummyStateStart struct {
+	canSend *atomic.Bool
+}
+
+var (
+	hexDecode = func(keyStr string) []byte {
+		keyBytes, _ := hex.DecodeString(keyStr)
+		return keyBytes
+	}
+	pub1, _ = btcec.ParsePubKey(hexDecode(
+		"02ec95e4e8ad994861b95fc5986eedaac24739e5ea3d0634db4c8ccd44cd" +
+			"a126ea",
+	))
+	pub2, _ = btcec.ParsePubKey(hexDecode(
+		"0356167ba3e54ac542e86e906d4186aba9ca0b9df45001c62b753d33fe06" +
+			"f5b4e8",
+	))
+)
+
+func (d *dummyStateStart) ProcessEvent(event dummyEvents, env *dummyEnv,
+) (*StateTransition[dummyEvents, *dummyEnv], error) {
+
+	switch event.(type) {
+	case *goToFin:
+		return &StateTransition[dummyEvents, *dummyEnv]{
+			NextState: &dummyStateFin{},
+		}, nil
+
+	// This state will loop back upon itself, but will also emit an event
+	// to head to the terminal state.
+	case *emitInternal:
+		return &StateTransition[dummyEvents, *dummyEnv]{
+			NextState: &dummyStateStart{},
+			NewEvents: fn.Some(EmittedEvent[dummyEvents]{
+				InternalEvent: fn.Some(dummyEvents(&goToFin{})),
+			}),
+		}, nil
+
+	// This state will proceed to the terminal state, but will emit all the
+	// possible daemon events.
+	case *daemonEvents:
+		// This send event can only succeed once the bool turns to
+		// true. After that, then we'll expect another event to take us
+		// to the final state.
+		sendEvent := &SendMsgEvent[dummyEvents]{
+			TargetPeer: *pub1,
+			SendWhen: fn.Some(func() bool {
+				return d.canSend.Load()
+			}),
+			PostSendEvent: fn.Some(dummyEvents(&goToFin{})),
+		}
+
+		// We'll also send out a normal send event that doesn't have
+		// any preconditions.
+		sendEvent2 := &SendMsgEvent[dummyEvents]{
+			TargetPeer: *pub2,
+		}
+
+		return &StateTransition[dummyEvents, *dummyEnv]{
+			// We'll state in this state until the send succeeds
+			// based on our predicate. Then it'll transition to the
+			// final state.
+			NextState: &dummyStateStart{
+				canSend: d.canSend,
+			},
+			NewEvents: fn.Some(EmittedEvent[dummyEvents]{
+				ExternalEvents: fn.Some(DaemonEventSet{
+					sendEvent, &DisableChannelEvent{},
+					&BroadcastTxn{}, sendEvent2,
+				}),
+			}),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unknown event: %T", event)
+}
+
+func (d *dummyStateStart) IsTerminal() bool {
+	return false
+}
+
+type dummyStateFin struct {
+}
+
+func (d *dummyStateFin) ProcessEvent(event dummyEvents, env *dummyEnv,
```

Comment:
I know you prefer this way of formatting long function definitions now. But can you please update our [formatting guide](https://github.com/lightningnetwork/lnd/blob/master/docs/code_formatting_rules.md#wrapping-long-function-definitions) which currently explicitly declares this as being wrong? Preferably with some rationale of when to use this over our previously preferred style:

```
func (d *dummyStateFin) ProcessEvent(event dummyEvents,
	env *dummyEnv) (*StateTransition[dummyEvents, *dummyEnv], error) {

```

Reply:
Yeah we def need a new rules for all the `[` stuff as well. I think I'm in the minority here re the aesthetic advantage of this style of wrapping, so will revert to match the style guide.

Reply:
I agree, we likely need an update to our code style to accommodate for the potentially longer types when using generic.
Just to show you what my main issue with the visual flow of this style is:

![code-style](https://github.com/lightningnetwork/lnd/assets/1008879/244db2c1-3a8c-46ca-ad7e-7e25d580b43a)

So I guess I would have less of a "problem" parsing this visual style instead:

```
func (d *dummyStateFin) ProcessEvent(event dummyEvents, env *dummyEnv,
	) (*StateTransition[dummyEvents, *dummyEnv], error) {
```

Tiny difference I guess, but would likely also be easier to describe in the style guide. What do you think?
But perhaps we should discuss in a dedicated issue/PR instead of here... Sorry to keep bringing it up :sweat_smile:

Reply:
Btw, with "safe space" I mean that things are always indented after the `func` keyword and the indentation only ends when the function ends.